### ISSUE TITLE
Add Factory Function Constructor to AutoDataAttribute

### ISSUE THREAD

**Issue by alexfoxgill** [AutoFixture/AutoFixture#358]
As requested by @moodmosaic in [this PR](https://github.com/AutoFixture/AutoFixture/pull/356#issuecomment-76664547)

As part of #356, the `AutoDataAttribute` needs to be able to provide data repeatedly to an xUnit theory. The "repeatedly" part is covered by #357 `InfiniteDataAtttribute`, which can wrap any `DataAttribute` to provide infinite repetitions of data.

However, in its current form, `AutoDataAttribute` is not completely suitable for this purpose. This is because each instance of the attribute takes a single **instance** of `IFixture`, which it uses to provide data to the xUnit theory method. This means that repeated iterations will re-use the fixture, potentially leading to unwanted side-effects with frozen objects.

This can be resolved by adding a factory function constructor to `AutoDataAttribute`:

``` CSharp
public class AutoDataAttribute
{
    public AutoDataAttribute(Func<IFixture> fixtureFactory)
    {
        this.fixtureFactory = fixtureFactory;
    }

    public AutoDataAttribute(Type fixtureType)
        : this(() => AutoDataAttribute.CreateFixture(fixtureType))
    {
    }

    public AutoDataAttribute(IFixture fixture)
        : this(() => fixture)
    {
    }

    ...
}
```

With a factory function, the fixture can be created on every iteration.

The only issue I can foresee with this approach is the public `Fixture` property on the attribute that permits inspection of the private field:

``` CSharp
public IFixture Fixture
{
    get { return this.fixture; }
}
```

With the factory approach there would be no `this.fixture` - rather, the property would have to return `this.fixtureFactory()`. I don't think that there should be a huge problem with this, as long as it is documented that the property will return an instance from a factory. Perhaps it could be renamed to `GetFixture()` (or removed entirely? Is it needed?)

**Comment by ploeh** [comment]
Adding a capability to repeat an `AutoData`-driven test multiple times is something I've been considering for years, but I've never been able to make up my mind whether this would be a desirable direction to go. That would take AutoFixture even closer towards Property-Based Testing, but still without shrinking.

Currently, when I want to do Property-Based Testing, I use [FsCheck](https://fscheck.github.io/FsCheck/) instead of AutoFixture.

Even if someone else would find this addition valuable, I don't want to inject a delegate into the constructor. For historic reasons, you can find a few examples of this approach to Constructor Injection here and there in the AutoFixture code base, but it's not a design principle I want to follow.

Delegates are fine for Method Injection, but for Constructor Injection, I prefer interfaces.

I'm a little wary of introducing yet another interface in the AutoFixture core library, but perhaps we could pay that tax only once by introducing an `IFactory<T>` generic interface.

**Comment by zvirja** [comment]
As far as I understand, this issue is a pre-cursor for the #357 and doesn't bring any value without that feature. Given that we closed #357, let's close this one.

@alexfoxgill Let me know if I misunderstood that.
